### DISCUSSION TITLE
Completed task property vs task checkbox

### DISCUSSION THREAD

**Discussion by mcastillof** [TechbeeAT/jtxBoard#172] [Q&A]
I was editing a task that I have completed  a few hours ago. I put in it its completed date and its status property as Completed.
After a while, I see that it was still marked as due in the tasklist. Only after ticking the checkbox (in the tasklist or in the task edit) it was truly set as completed.

Is this the expected behavior? If it is, then probably a text in the checkbox of the task edit view could help.

I was thinking that if I put its property as completed, then its due status would have been removed. I also thought that if a completed date was set, then the due status would have been removed.

And now that I'm typing this, I think that if you set the status of a task as completed, or if you tick the checkbox, the completed date property should be set to the current time if it wasn't set previously. if it was set previously, then it is also a good idea to also tick the checkbox and set the status property as completed. 

what do you think?

**Comment by patrickunterwegs** [comment]
Hey @mcastillof , thanks for that!
At the very beginning of the app those fields were already linked, but then there was a complaint that this is not defined/intended by the standard. Currently the status is updated on progress change when the status was updated at least once. The completed date is not set automatically.
This is indeed a bit problematic, there is actually no clear definition when a task is really completed - when the progress is set to 100 or when the status is changed to completed or both? All other systems that I checked used the progress field to mark tasks as completed. 

However, I am absolutely with you and I suggest a setting to "link progress, status and completed" with a bit of a logic. 
If the option is activated:
- new entries are created with status "Needs action" instead of no status
- when the status is changed
-- to "Needs action", the progress is set to 0 and the completed field is cleared
-- to "In process", the progress is kept (if it was between 1..99) or set to 1, the completed field is cleared
-- to "Completed", the progress is set to 100 and the completed field is set to now
-- to "Cancelled", no other field changes
- when the progress is changed to
-- 0, the status is set to "Needs action", the completed field is cleared
-- 1..99, the status is set to "In process, the completed field is cleared
-- 100, the status is set to "Completed", the completed field is set to now
- when the completed field is set to a date
-- the progress is set to 100, the status is set to "Completed"

If the option is deactivated, no logic is applied, it is up to the user to set the fields according to their needs. 

Would you also say that this is an approach that makes sense?

**Reply by mcastillof** [reply]
Sure @patrickunterwegs, your logic makes all sense. It is what I was expecting it to behave when I started to use the app a few days ago. 

Now, considering #173:
> -- to "Cancelled", no other field changes
maybe cancelled could behave just as a completed task. Meaning that if you set a task as completed or canceled it will not been shown in the calendar activity.

Also, I'm using the latest version of jtx Board installed from F-Droid. I don't see a progress UI control. Is it in the beta release?

**Reply by patrickunterwegs** [reply]
Well, yes, I agree with your expectation, still there are users out there who would start complaining if a logic is applied that is not defined in the standard 😉 I'm just working on that and I'll add a new setting where the logic can be deactivated (default will be activated).

The progress bar is another topic where users complained that it is bothering for their use cases. I made it optional, you can activate it in the settings!
